src/views/portfolio/projects/ProjectCreateProjectModal.vue

```diff
@@ -166,25 +182,55 @@
         }).then((response) => {
           this.$emit('refreshTable');
           this.$toastr.s(this.$t('message.project_created'));
+          this.selectedParent = null;
+          this.availableParents = [{ value: null, text: ''}]
         }).catch((error) => {
           this.$toastr.w(this.$t('condition.unsuccessful_action'));
         }).finally(() => {
           this.$root.$emit('bv::hide::modal', 'projectCreateProjectModal');
         });
       },
       retrieveLicenses: function() {
-        let url = `${this.$api.BASE_URL}/${this.$api.URL_LICENSE_CONCISE}`;
-        this.axios.get(url).then((response) => {
-          for (let i = 0; i < response.data.length; i++) {
-            let license = response.data[i];
-            this.selectableLicenses.push({value: license.licenseId, text: license.name});
-            if (this.project.resolvedLicense && this.project.resolvedLicense.uuid === license.uuid ) {
-              this.selectedLicense = license.licenseId;
+        return new Promise(resolve => {
+          let url = `${this.$api.BASE_URL}/${this.$api.URL_LICENSE_CONCISE}`;
+          this.axios.get(url).then((response) => {
+            for (let i = 0; i < response.data.length; i++) {
+              let license = response.data[i];
+              this.selectableLicenses.push({value: license.licenseId, text: license.name});
+              if (this.project.resolvedLicense && this.project.resolvedLicense.uuid === license.uuid ) {
+                this.selectedLicense = license.licenseId;
+              }
             }
-          }
-        }).catch((error) => {
-          this.$toastr.w(this.$t('condition.unsuccessful_action'));
-        });
+          }).catch((error) => {
+            this.$toastr.w(this.$t('condition.unsuccessful_action'));
+          }).finally(() => {
+            resolve()
+          });
+        })
+      },
+      retrieveParents: function() {
+        return new Promise(resolve => {
+          let url = `${this.$api.BASE_URL}/${this.$api.URL_PROJECT}`;
+          this.axios.get(url).then((response) => {
+            for (let i = 0; i < response.data.length; i++) {
+              let project = response.data[i];
+              if (project.active) {
+                if (project.version) {
+                  this.availableParents.push({value: project.uuid, text: project.name + ' : ' + project.version});
+                } else {
+                  this.availableParents.push({value: project.uuid, text: project.name});
+                }
+              }
+              if (this.project.parent && this.project.parent.uuid === project.uuid ) {
+                this.selectedParent = project.uuid;
+              }
+            }
+          }).catch((error) => {
+            this.$toastr.w(this.$t('condition.unsuccessful_action'));
+          }).finally(() => {
+            resolve()
+          });
+        })
```

Comment:
It just now dawned on me that this will either not load enough projects, or will be an immensely expensive operation in large portfolios.

The default `pageSize` for paginated APIs is 100. So if there are 105 projects, only the first 100 will be fetched here. You may get unlucky and the current parent is not among those 100 projects, causing the selected value to wrongly be empty.

If this limitation was not there, then we'd load too much data at once, and the dropdown would grow to an unmanageable size.

I'm wondering if maybe this should be a search field instead. The API server has a `/api/v1/search/project` endpoint that could be used:

https://github.com/DependencyTrack/dependency-track/blob/9c3290998e09c29ab4ffdb60cf1632ec4b72b07b/src/main/java/org/dependencytrack/resources/v1/SearchResource.java#L67-L83

Thoughts?

Reply:
Good question... After some quick googling, this looks promising: https://vue-multiselect.js.org/#sub-asynchronous-select

Not sure how well this blends in with the bootstrap-based UI though.

Reply:
I think that `vue-multiselect` would be a very good fit for this case and should blend in well with the already existing UI, if correctly adjusted.

However, I'm not quite sure about using the `/api/v1/search/project` endpoint for searching a parent project.
The endpoint uses `PROJECT_UUID`, `PROJECT_NAME`, `PROJECT_VERSION` and `PROJECT_DESCRIPTION` as search criteria (and can't be turned off, as far as I've seen it) for a project, but only `PROJECT_NAME` and `PROJECT_VERSION` are required in this dialogue, so the other criteria could match projects which wouldn't make sense for the user and create unnecessary results.

Another problem would be, that the endpoint only returns a projects UUID, name, version and description, but not its active status, which is essential for the dialogue because inactive projects cannot be selected as parents. The active status of a project could be added to the IndexConstants, but I don't know if this would cause any problems in other existing logic.

Reply:
I think you're right about the search endpoint. Even adding the `active` field would still cause more elements being returned by the API than the UI will actually display (filtering on client-side is kinda meh). And the current search endpoint will always perform a "contains" query, without users being able to specify which field to search in.

But you may be able to use the `/api/v1/project?searchText=<fieldInput>&excludeInactive=false` endpoint, which is also what the search field in the project list uses:

![Screenshot 2022-11-27 at 22 41 39](https://user-images.githubusercontent.com/5693141/204161133-ad9093ea-c744-40b2-9d97-7ab486a280bb.png)